File: Respuestas_Examen_Parcial/Pregunta1/Codigo1.py
# treap_augmented.py
from __future__ import annotations
from typing import Optional
import random

class TreapNode:
    """Nodo de Treap con máximo del subárbol aumentado."""
    __slots__ = ("key", "value", "priority", "left", "right", "subtree_max")

    def __init__(self, key: int, value: int, priority: float | None = None):
        self.key = key
        self.value = value
        self.priority = priority if priority is not None else random.random()
        self.left: Optional[TreapNode] = None
        self.right: Optional[TreapNode] = None
        self.subtree_max = value  # inicializa con su propio valor

    # ---------- funciones auxiliares ----------
    def _update(self) -> None:
        """Recalcula el máximo del subárbol después de un cambio estructural."""
        self.subtree_max = self.value
        if self.left:
            self.subtree_max = max(self.subtree_max, self.left.subtree_max)
        if self.right:
            self.subtree_max = max(self.subtree_max, self.right.subtree_max)

def rotate_left(p: TreapNode) -> TreapNode:
    """Retorna la nueva raíz después de una rotación a la izquierda."""
    r = p.right
    assert r is not None
    p.right = r.left
    r.left = p
    p._update()
    r._update()
    return r

def rotate_right(p: TreapNode) -> TreapNode:
    """Retorna la nueva raíz después de una rotación a la derecha."""
    l = p.left
    assert l is not None
    p.left = l.right
    l.right = p
    p._update()
    l._update()
    return l

# ---------- inserción ----------
def insert(root: Optional[TreapNode], key: int, value: int,
           priority: float | None = None) -> TreapNode:
    if root is None:
        return TreapNode(key, value, priority)

    if key < root.key:
        root.left = insert(root.left, key, value, priority)
        if root.left.priority < root.priority:          # propiedad de min-heap
            root = rotate_right(root)
    elif key > root.key:
        root.right = insert(root.right, key, value, priority)
        if root.right.priority < root.priority:
            root = rotate_left(root)
    else:  # sobrescribe el valor si la clave ya existe (claves únicas)
        root.value = value
    root._update()
    return root

# ---------- eliminación ----------
def delete(root: Optional[TreapNode], key: int) -> Optional[TreapNode]:
    if root is None:
        return None
    if key < root.key:
        root.left = delete(root.left, key)
    elif key > root.key:
        root.right = delete(root.right, key)
    else:  # se encontró el nodo a eliminar
        if not (root.left or root.right):  # hoja
            return None
        # rota el hijo con menor prioridad hacia arriba
        if root.left and (not root.right or root.left.priority < root.right.priority):
            root = rotate_right(root)
            root.right = delete(root.right, key)
        else:
            root = rotate_left(root)
            root.left = delete(root.left, key)
    root._update()
    return root

# ---------- consulta de máximo en rango ----------
def find_max_value_in_range(root: Optional[TreapNode],
                            key_min: int, key_max: int) -> int | None:
    """Retorna el máximo valor para claves en el rango [key_min, key_max]."""
    if root is None or key_min > key_max:
        return None

    max_val = None
    node = root
    stack: list[TreapNode] = []
    while stack or node:
        while node:
            # Podar todo el subárbol si no puede contener claves en el rango
            if node.key < key_min:
                node = node.right
            elif node.key > key_max:
                node = node.left
            else:
                stack.append(node)
                node = node.left
        node = stack.pop()
        # node.key está garantizado dentro del rango
        if max_val is None or node.value > max_val:
            max_val = node.value
        # Si todo el subárbol derecho está dentro del rango, se puede usar el valor aumentado
        if node.right and node.right.key <= key_max:
            max_val = max(max_val, node.right.subtree_max)
            node = None  # terminado
        else:
            node = node.right
    return max_val

File: Respuestas_Examen_Parcial/Pregunta1/Codigo1.cpp
#include <iostream>      // std::cout
#include <memory>        // std::unique_ptr, std::make_unique
#include <optional>      // std::optional, std::nullopt
#include <random>        // std::mt19937, std::uniform_real_distribution
#include <limits>        // std::numeric_limits
#include <algorithm>     // std::max
#include <utility>       // std::move

struct TreapNode {
    int key;
    int value;
    double priority;
    std::unique_ptr<TreapNode> left;
    std::unique_ptr<TreapNode> right;
    int subtree_max;

    TreapNode(int k, int v, std::optional<double> p = std::nullopt)
        : key{k}, value{v},
          priority{p.value_or(random_priority())},
          subtree_max{v} {}

private:
    static double random_priority() {
        static thread_local std::mt19937 rng(std::random_device{}());
        static std::uniform_real_distribution<double> dist(0.0, 1.0);
        return dist(rng);
    }
};

/* ---------- utilidades ---------- */
static void update(TreapNode* n) {
    if (!n) return;
    n->subtree_max = n->value;
    if (n->left)  n->subtree_max = std::max(n->subtree_max, n->left->subtree_max);
    if (n->right) n->subtree_max = std::max(n->subtree_max, n->right->subtree_max);
}

static std::unique_ptr<TreapNode> rotate_left(std::unique_ptr<TreapNode> p) {
    auto r = std::move(p->right);
    p->right = std::move(r->left);
    r->left = std::move(p);
    update(r->left.get());
    update(r.get());
    return r;
}

static std::unique_ptr<TreapNode> rotate_right(std::unique_ptr<TreapNode> p) {
    auto l = std::move(p->left);
    p->left = std::move(l->right);
    l->right = std::move(p);
    update(l->right.get());
    update(l.get());
    return l;
}

/* ---------- inserción ---------- */
std::unique_ptr<TreapNode> insert(std::unique_ptr<TreapNode> root,
                                  int key, int value,
                                  std::optional<double> pr = std::nullopt) {
    if (!root) return std::make_unique<TreapNode>(key, value, pr);

    if (key < root->key) {
        root->left = insert(std::move(root->left), key, value, pr);
        if (root->left->priority < root->priority)
            root = rotate_right(std::move(root));
    } else if (key > root->key) {
        root->right = insert(std::move(root->right), key, value, pr);
        if (root->right->priority < root->priority)
            root = rotate_left(std::move(root));
    } else {
        root->value = value;            // clave existente → actualiza valor
    }
    update(root.get());
    return root;
}

/* ---------- eliminación ---------- */
std::unique_ptr<TreapNode> delete_key(std::unique_ptr<TreapNode> root, int key) {
    if (!root) return nullptr;
    if (key < root->key) {
        root->left = delete_key(std::move(root->left), key);
    } else if (key > root->key) {
        root->right = delete_key(std::move(root->right), key);
    } else {
        if (!root->left && !root->right) return nullptr;
        if (root->left &&
            (!root->right || root->left->priority < root->right->priority)) {
            root = rotate_right(std::move(root));
            root->right = delete_key(std::move(root->right), key);
        } else {
            root = rotate_left(std::move(root));
            root->left = delete_key(std::move(root->left), key);
        }
    }
    update(root.get());
    return root;
}

/* ---------- consulta de rango ---------- */
std::optional<int> max_in_range(const TreapNode* root, int lo, int hi) {
    if (!root || lo > hi) return std::nullopt;
    const TreapNode* cur = root;
    int best = std::numeric_limits<int>::min();

    while (cur) {
        if (cur->key < lo) {
            cur = cur->right.get();
        } else if (cur->key > hi) {
            cur = cur->left.get();
        } else {
            best = std::max(best, cur->value);
            if (cur->left && cur->left->key >= lo)
                best = std::max(best, cur->left->subtree_max);
            cur = cur->right.get();
        }
    }
    return (best == std::numeric_limits<int>::min())
           ? std::nullopt
           : std::optional<int>(best);
}

/* ---------- demo mínima ---------- */
int main() {
    std::unique_ptr<TreapNode> root;
    root = insert(std::move(root), 10, 42);
    root = insert(std::move(root), 5, 99);
    root = insert(std::move(root), 15, 7);

    auto mx = max_in_range(root.get(), 1, 20);
    if (mx) std::cout << "Valor máximo en [1,20] = " << *mx << '\n';
    return 0;
}
